Support marshalling of bigarrays with dimensions that don't fit in 4 bytes

yallop opened the pull request:
The bigarray code has a comment justifying the use of 4-byte integers for marshalling dimensions:

```c
  /* On a 64-bit machine, if any of the dimensions is >= 2^32,
     the size of the marshaled data will be >= 2^32 and
     extern_value() will fail.  So, it is safe to write the dimensions
     as 32-bit unsigned integers. */
```

However, since #224 (*Marshaling: beyond 4 Gb*) the comment is no longer true.

This PR changes the marshalling format for bigarrays (in a backwards-incompatible way) to use 8-byte integers to avoid overflow/truncation.  There's also a test (9526f23) that fails before the change and passes afterwards.

yallop committed:
Add a test for the variable-length bigarray dimension marshalling scheme

yallop committed:
Use @stedolan's simple variable-length encoding for bigarray dimensions

xavierleroy commented:
I'm not happy with the increase in size of marshaled data that will penalize all users while a tiny fraction will benefit from the change.  Especially since bigarrays are increasingly used for what they were not designed for, namely small, unidimensional arrays (see e.g. cstruct).

What about variable-length encoding of the dimensions?  We'll be changing the marshaled data format anyway, so there is room for a good design.

xavierleroy reviewed (changes requested):
Since my review was requested: my review is negative.

yallop commented:
> What about variable-length encoding of the dimensions? We'll be changing the marshaled data format anyway, so there is room for a good design.

Okay, I'll look at switching to a variable-length encoding.  Let me know if you have a preference for a particular approach.  If not, I'll probably implement a UTF-8-style scheme.

xavierleroy commented:
I was thinking of BER from ASN/1:
- `0xxxxxxx` for integers between 0 and 127.
-  Otherwise, first byte is `1nnnnnnn`, where the 7 bits `nnnnnnn` represent the number N of bytes, followed by N bytes (in big endian).

Since we know that 1 <= N <= 8 always, we can use 3 bits to encode the length N - 1, and 4 bits for data:
- `0xxxxxxx`
- `1000xxxx` `xxxxxxxx`
- `1001xxxx` `xxxxxxxx`
- `1010xxxx` `xxxxxxxx`  `xxxxxxxx`
- etc.

xavierleroy commented:
Another encoding, probably easier to implement with the I/O functions we already have:
* `0xxxxxxx` for numbers between 0 and 0x7F
* `1xxxxxxx xxxxxxxx` for numbers up to 0x7DFF  (hence first byte is not 0xFE nor 0xFF)
* `0xFE` followed by 4 bytes for numbers up to 2^32 - 1
* `0xFF` followed by 8 bytes for numbers up to 2^64 - 1

mshinwell commented:
Another possibility might be uleb128 as used by DWARF.  I can provide conversion functions for that.

stedolan commented:
These are length fields, so space efficiency can only matter when the length is short. (Otherwise, the overhead is a tiny fraction of the space used by the actual data, regardless of encoding). So, how about something simpler, like:

encode:
```c
    if (len < 0xffff) {
      caml_serialize_int_2(len);
    } else {
      caml_serialize_int_2(0xffff);
      caml_serialize_int_8(len);
    }
```
decode:
```c
    len = caml_deserialize_uint_2();
    if (len == 0xffff) len = caml_deserialize_uint_8();
```

For short lengths, this is half the size of the current 4-byte field. For long lengths, the worst-case overhead is 10 bytes per 64k, around 0.02%.

yallop committed:
Changes entry: use variable-length encoding to marshal bigarray dimensions

yallop commented:
I've rebased on trunk and switched to @stedolan's [simple variable-length scheme](https://github.com/ocaml/ocaml/pull/8791#issuecomment-509202370).

xavierleroy reviewed (approved):
Looks good to me, thanks.

The only question I have is whether we should change the identifier for marshalled bigarrays:
https://github.com/ocaml/ocaml/blob/ec24f98bec9c62da5f7e1cd1b1495fbd0f0c50e8/runtime/bigarray.c#L70

This would protect against programs that saved a bigarray on disk with the old format, then are recompiled with a newer OCaml version, then try to read the bigarray with the new format.

On the one hand we've never guaranteed that this should work.

On the other hand it would be trivial to change the identifier from `_bigarray` to `_bigarr02` (say), basically using two bytes of the identifier for a version number.

mshinwell commented:
@xavierleroy I would be in favour of bumping the magic number, given that I've seen some `Bigarray` structures being marshalled recently, in the context of the `Owl` scientific library.

yallop committed:
Change the bigarray identifier: _bigarray ~> _bigarr02

The identifier change reflects a change in the bigarray marshalling
format to use a variable-length encoding for dimensions.

yallop commented:
I've updated the `"_bigarray"` identifier to `"_bigarr02"` (b1f81a1a) and rebased.

xavierleroy reviewed (approved):
Thanks for changing the marshalling identifier.  All remarks were addressed, so I'll merge ASAP.

xavierleroy commented:
Cheery-picked to 4.11 because this PR can be viewed as a bug fix, and because why not.
